dht_rt_seeder.go

```diff
@@ -0,0 +1,111 @@
+package dht
+
+import (
+	"context"
+
+	"github.com/ipfs/go-todocounter"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+func (dht *IpfsDHT) seedRoutingTable(candidates, fallbacks []peer.ID) error {
+	peersFromSet := func(m map[peer.ID]struct{}) []peer.ID {
+		peers := make([]peer.ID, len(m))
+		i := 0
+		for p := range m {
+			peers[i] = p
+			i++
+		}
+
+		return peers
+	}
+
+	// candidates we have still not tried dialling to
+	remainingCandidates := make(map[peer.ID]struct{})
+	// fallbacks we have still not tried dialling to
+	remainingFallbacks := make(map[peer.ID]struct{})
+
+	for i := range candidates {
+		remainingCandidates[candidates[i]] = struct{}{}
+	}
+
+	for i := range fallbacks {
+		remainingFallbacks[fallbacks[i]] = struct{}{}
+	}
+
+	for {
+
+		todial := dht.seedsProposer.Propose(dht.routingTable, peersFromSet(remainingCandidates), peersFromSet(remainingFallbacks))
+		if len(todial) == 0 {
+			logger.Infof("dht rt seeder: finished as proposer returned nil, RT size is now %d", dht.routingTable.Size())
+			return nil
+		} else {
+			// start dialing
+			type result struct {
+				p   peer.ID
+				err error
+			}
+
+			// attempts to dial to a given peer to verify it's available
+			dialFn := func(ctx context.Context, p peer.ID, res chan<- result) {
```

Comment:
should be able to define this outside of the for loop

Reply:
Problem has been fixed.